### ISSUE TITLE
RegisterWaitForSingleObject deadlocks on Unix

### ISSUE THREAD

**Issue by stephentoub** [dotnet/coreclr#2093]
Not as simple a repro as I'd like, but here's my repro:
```C#
using System; 
using System.Threading;
using System.Threading.Tasks;

class Test
{
    static void Main()
    {
        for (int i = 0; ; i++)
        {
            Console.WriteLine(i);
            Task.WaitAll(Task.Run((Action)Repro), Task.Run((Action)Repro), Task.Run((Action)Repro), Task.Run((Action)Repro));
        }
    }

    static void Repro()
    {
        var syncObj = new object();
        var signalCallback = new ManualResetEvent(false);
        var callbackDone = new ManualResetEvent(false);

        RegisteredWaitHandle rwh = null;
        rwh = ThreadPool.RegisterWaitForSingleObject(signalCallback, (_, __) =>
        {
            lock (syncObj) { rwh.Unregister(null); }
            callbackDone.Set();
        }, null, -1, true);

        // Queue a work item that'll complete the signal callback       
        Task.Run(() => { signalCallback.Set(); });

        // Create some other dummy work
        var tasks = new Task[Environment.ProcessorCount - 1];
        for (int i = 0; i < tasks.Length; i++)
        {
            tasks[i] = Task.Run(() => { lock (syncObj) { for (int j=0; j<1000; j++){} } });
        }

        // Wait for the callback to be invoked and unregister
        callbackDone.WaitOne();
        Task.WaitAll(tasks);
    }
}
```
On Windows, both coreclr and desktop, this continues to loop in the Main method successfully.  On Linux after one or a few iterations, it stops iterating, and the debugger shows a thread blocked waiting on an event inside of the Unregister method.

This is similar to code in System.Diagnostics.Process, and it appears to be causing some Process.Exited events to not be raised.

**Comment by janvorli** [comment]
Fixed by #2098
CC: @stephentoub
